### DISCUSSION TITLE
認証レベル以外のHub連携でクラッシュ

### DISCUSSION THREAD

**Discussion by o-hkmr** [pixiv/vroid-sdk-developers#189] [Q&A]
VRoid SDK 1.31を利用し、以下のURLを参考にBP_VRoidMannequinRetargetCharacterを使用した任意のLevelにBP_VRoidRuntimeAuthActorを配置し、キャラクターを選択できるようにしました。
https://vroid.notion.site/VRoid-SDK-for-Unreal-Engine-21321b5e31b247e99600f270b3414d4e#fbcc958168e44da397e933e28e0f5575

キャラクターを選んだ後に成功する場合もあれば、ロード中にクラッシュする場合もあり、パッケージ化した場合でも同様にクラッシュするケースがあります。一度成功しても２回同じ操作（ウィジェット表示してキャラクター選択）をするとほぼクラッシュします。
![Clush](https://github.com/user-attachments/assets/18d31111-f0b5-4385-90e4-08040583ed9e)

一度読み込めたモデルをアプリの再起動後に読み込むと成功する確率が高い印象です。

UnrealEngineは5.5を利用しており、VRoid SDK 1.31は5.4までしかサポートしていませんが、ビルドも成功し基本的な動作に影響はないのではと考えています。

サポートいただけると幸いです。

クラッシュ時のログを添付します。
[TestApp.log](https://github.com/user-attachments/files/18272148/TestApp.log)

**Comment by mameoPix** [comment]
VRoid SDK 1.3.1 + UE5.5の組み合わせにてクラッシュするとの報告をいただき、ありがとうございます。
UE5.5をまだサポート対象に含めることが出来ていない要因として[レンダリングの並列化](
https://dev.epicgames.com/documentation/ja-jp/unreal-engine/unreal-engine-5.5-release-notes#%E3%83%AC%E3%83%B3%E3%83%80%E3%83%AA%E3%83%B3%E3%82%B0%E3%81%AE%E4%B8%A6%E5%88%97%E5%8C%96)に対応出来ていないことが大きいです。

そこで、試していただきたいことが3点あります。
**1. レンダリングスレッドの無効化**
　　プロジェクトの`Config/DefaultEngine.ini`を開いて以下を追記してみてください。
```
[RenderingThread]
bAllowThreadedRendering=False
```
　　SDKの開発環境でもまだ動作テスト中なのですが、UE5.5のレンダリングの並列化を無効化出来る可能性があります。
　　ただし、UE5.5の恩恵であるレンダリングのパフォーマンス向上が無効化されてしまうことになります。
　　そのため、現在この設定を行わずとも動作するように調整中です。


**2. VRM4Uのコードに追記する (要 C++ビルド環境)**
　　この変更はパッケージ化した際の挙動にのみ影響します。そのため、Editor環境での動作には影響しません。
　　`Plugins\VRM4U\Source\VRM4ULoader\Private\VrmConvertModel.cpp`内の**1937行目**辺りに以下のコードがあります。
```
	d.SkinWeightVertexBuffer = InWeights;
}
#endif	// editor
```
　　このコードを以下のように更新してください。
```
	d.SkinWeightVertexBuffer = InWeights;
	if (d.SkinWeightVertexBuffer.GetDataVertexBuffer() != nullptr)
	{
		d.SkinWeightVertexBuffer.GetDataVertexBuffer()->UpdateRHI(RHICmdList);
	}
#endif	// editor
```
　　この処理を追加することでパッケージ化した際のスキンウェイトの頂点Bufferの更新を明示的に行うようになります。
　　コードを修正後、ビルドし直してVRM4UのDLLを更新する必要があります。
　　添付いただいた画像のVertexBufferのエラーを直接回避出来る可能性があります。


**3. 1. 2. を試してもダメだった場合**
　　1. の時と同様にDefaultEngine.iniを開いて、以下を追記してください。
```
[Rendering]
r.EnableMultiThreadedRendering=0

[RenderingThread]
bAllowAsyncRenderThreadUpdates=False
bAllowThreadedRendering=False
```
　　1.で行った変更と意味合いが重複しているものもありますが
　　この設定変更は本来デバッグ用に一時的にレンダリングの並列化を無効化するものです。
　　そのため、パフォーマンスが著しく低下することが予想されます。


VRoid SDK 1.4.0 のリリース時にはUE5.5サポートを含めてより安定した動作を行えるように調整して参ります。
また追加で質問がある場合や、この対応でも動作が改善しない場合はコメントにて追記いただけますと幸いです。

**Reply by o-hkmr** [reply]
早速のサポートありがとうございます。
「1. レンダリングスレッドの無効化」においては効果がありませんでした。
「2. VRM4Uのコードに追記する 」を実施したところ、パッケージ化したイメージでは繰り返しキャラクターを変更してもほとんどクラッシュしなくなりました。ありがとうございます！
ほとんど、というのはパッケージ化したイメージで１度だけクラッシュしたのですが、それ以降は今のところクラッシュしていません。

アプリはUE5.5で開発しているため、未サポートと知りつつVRoid SDK 1.3.1を使用させていただいています。
ひとまずこの状態で開発継続できるので助かりました。

できればVRoid SDK1.4.0のリリース予定時期を教えていただけると幸いです。

**Reply by PaperSloth** [reply]
1.4.0のリリース時期について、改めておおよその時期が分かり次第共有させていただきます🙇‍♂️

**Reply by mameoPix** [reply]
回答をお待たせしました。
1.4.0のリリース予定ですが、今月末から来月上旬頃を予定してリリースに向けた調整を行っております。
もうしばらくお待ちいただけますと幸いです。

**Comment by o-hkmr** [comment]
わざわざご回答いただきありがとうございます。
楽しみに待っていたいと思います。
